### ISSUE TITLE
Fragment spreads on a union type cause an error

### ISSUE THREAD

**Issue by tberman** [gql-dart/gql#228]
Due to the way fragment spreads are handled (by walking them and including them in the parent selection set) when you have a fragment spread on a union you get an error, for example:

```
query SearchQuery {
  search {
    ...DroidFragment
  }
}

fragment DroidFragment on Droid {
  name
}
```

causes the error

```
Exception: SearchResult is not an ObjectTypeDefinitionNode or InterfaceTypeDefinitionNode
```

It appears that fragment spreads should be handled more like inline fragments and have built subclasses, but I am likely missing something.

Happy to attempt to put together a PR if there is consensus on what the right behavior would be!

**Comment by agent3bood** [comment]
Hi

Would you please share the schema, or provide a complete example to try to reproduce it.
I have used Fragments and it is just the way you did it.

**Comment by tberman** [comment]
The example provided will fail against the swapi schema used in the `end_to_end` examples in this repo. I can push up a branch that shows the failure. It's specifically from fragment spreads on unions not wrapped inside an inline fragment. You can see in some other examples in `end_to_end` where the spread is being wrapped in an inline fragment to work around this.

**Comment by tberman** [comment]
Pushed up a [branch](https://github.com/gql-dart/gql/compare/master...tberman:union_test_case) that shows the issue.

If you run build, you will get the error I mentioned.

I did some quick digging and if you look at how https://github.com/gql-dart/gql/blob/master/codegen/gql_code_builder/lib/src/operation/data.dart#L124 handles FragmentSpreads it hoists them onto the parent, causing this (correct) error for unions. Instead, it seems like it should be handled as the function handles InlineFragmentNodes: https://github.com/gql-dart/gql/blob/master/codegen/gql_code_builder/lib/src/operation/data.dart#L160

**Comment by tberman** [comment]
@agent3bood I am happy to spend some time poking at this if we get to a place where there is some general agreement on the right solution!

**Comment by agent3bood** [comment]
Hello again 

First I want to point to some resources, where in both documentations the way to select a union is to use inline fragment, which is working fine. 

https://graphql.org/learn/queries/#inline-fragments
https://www.apollographql.com/docs/android/essentials/inline-fragments/




*working as per standards*
```
query Search {
    search {
        ... on Human {
            name
        }
        ... on Droid {
            ...DroidFragment
        }
    }
}

fragment DroidFragment on Droid {
    name
}
```




I am not sure wether the suggested way is a standard or a `shorthand` developed by the community.

*not working*

```
query Search {
    search {
        ... on Human {
            name
        }
        ...DroidFragment # <<< causing error
    }
}

fragment DroidFragment on Droid {
    name
}
```



Please join the [Discord channel](https://discord.gg/ySTwQzSr) to provide your solution for this issue, and possibly get it merged. 


Thanks for taking the time to report this issue.

**Comment by tberman** [comment]
I will join the Discord channel to chat more about a solution!

As far as the query its definitely standard, if you look at the spec linked, you can see the line:

```
Named fragments can also be used in the same way, since a named fragment always has a type attached.
```

Additionally, its supported very clearly by the AST and validators.

**Comment by agent3bood** [comment]
>  Named fragments can also be used in the same way

But I think you still need to add the `... on`  like bellow

```
query Search {
    search {
# inline fragment 
        ... on Human {
            name
        }
# named fragment 
        ... on Droid {
            ...DroidFragment
        }
    }
}

fragment DroidFragment on Droid {
    name
}
```

**Comment by smkhalsa** [comment]
Per the [spec](https://spec.graphql.org/draft/#sec-Abstract-Spreads-in-Abstract-Scope.Interface-Spreads-in-implemented-Interface-Scope), this should be possible. PRs to add this functionality welcome.

**Comment by smkhalsa** [comment]
Also, I ran into an issue a while back where the builder was failing when using inline fragments inside a fragment on an interface, and I expect it is a related issue.

**Comment by smkhalsa** [comment]
@tberman did you ever work up a solution for this?

**Comment by tberman** [comment]
No, I pinged in discord about it as asked, didn't get any engagement. Happy to discuss async here as well if thats better, but would like to get to some agreement on what the shape of a solution would look like before diving in.

A quick glance made me think the fragment spreads should be treated the same as the inline fragments instead of how they are now, which would kind of mask them out a bit.

What do you think?

**Comment by smkhalsa** [comment]
> Happy to discuss async here as well if thats better, but would like to get to some agreement on what the shape of a solution would look like before diving in.

Sure

> fragment spreads should be treated the same as the inline fragments

Can you expand on this? I haven't looked at this section of the code in a while.

**Comment by 1awaleed** [comment]
I am also having the same issue here.
